### ISSUE TITLE
Completion based on trait bounds

### ISSUE THREAD

**Issue by kngwyu** [racer-rust/racer#939]
Related: #706, #829 
## General roadmap
- [x] for `FnArg`
already implemented in #828, #937 
```rust
fn f<T: Clone>(a: T) {
    let b = a.clo~
}
```
- [ ] for struct member
```rust
struct S<T> {
    mem: T
}
fn f<T: Clone>(a: S<T>) {
    let b = a.mem.clo~
}
impl<T: Clone> S<T> {
    fn f(self) {
        let b = a.mem.clo~
    } 
}
```
- [ ] for closures
```rust
fn fun<F: Fn() -> Option<i64>>(f: F) {
    let a = f().unwra~
}
```

### refactor generics
- [x] Intoroduce ImplHeader 
- [ ] Replace `MatchType::Impl` and `MatchType::TraitImpl` with ImplHeader
- [ ] Remove `Match::generics_args` and `Match::generic_types`
- [ ] Modify `MatchType::Enum` and etc to have generics
- [ ] Intoroduce `MatchType::Method` which has ImplHeader

### enhance `scope_start`'s inefficiency
- [ ] Have scope cache

### Method search
- [ ] rewrite method search using ImplHeader

### TypeName resolution
- [ ] resolve type parameter in ImplHeader when `resolve_path` is called

**Comment by letmutx** [comment]
Hi! Is struct member method completions working? I can't seem to get completions with the latest master. I am using the same example above
```
struct S<T> {
    mem: T
}
fn f<T: Clone>(a: S<T>) {
    let b = a.mem.clo~
}
```
```
$ RUST_LOG=racer=trace ../racer/target/debug/racer complete 5 21 src/main.rs 2> trace.txt
PREFIX 71,74,clo
END
```
[trace.txt](https://github.com/racer-rust/racer/files/2484934/trace.txt)

Weirdly enough, the testcases in `tests/trait_bounds.rs` pass.

**Comment by kngwyu** [comment]
Thanks. 

> Weirdly enough, the testcases in tests/trait_bounds.rs pass.

Tests contain only 'bounded by impl' cases, like
```rust
impl<T: Clone> S<T> {
    fn f(self) {
        let b = a.mem.clo~
    } 
}
```
, so it's not weird.
I don't remember it's simply not implemented or broken, but anyway fix it later.

**Comment by letmutx** [comment]
Right. I was obviously overlooking that. Sorry. I would like to take a go
at this.

**Comment by kngwyu** [comment]
@letmutx 
Thanks, but I want to fix this by myself, to recall what I implemented in last month.

**Comment by letmutx** [comment]
Okay. I have some progress here: https://github.com/letmutx/racer/tree/add-struct-member-trait-completions
